core/embed/prodtest/main.c

```diff
@@ -513,9 +603,443 @@ static void test_otp_write_device_variant(const char *args) {
+  }
+
+  // Verify whether the secret was stored correctly in flash and OPTIGA.
+  memzero(secret, sizeof(secret));
+  if (secret_read(secret, SECRET_OPTIGA_KEY_OFFSET, SECRET_OPTIGA_KEY_LEN) !=
+      sectrue) {
+    vcp_println("ERROR Failed to read pairing secret.");
+    return false;
+  }
+  */
+
+  ret = optiga_sec_chan_handshake(secret, sizeof(secret));
+  memzero(secret, sizeof(secret));
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_sec_chan_handshake error %d.", ret);
+    return false;
+  }
+
+  return true;
+}
+
+static void optiga_lock(void) {
+  if (!pair_optiga()) {
+    return;
+  }
+
+  // Delete trust anchor.
+  optiga_result ret =
+      optiga_set_data_object(0xe0e8, false, (const uint8_t *)"\0", 1);
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_set_data error %d for 0xe0e8.", ret);
+    return;
+  }
+
+  // Set data object metadata.
+  static const optiga_metadata_item ACCESS_PAIRED = {
+      (const uint8_t *)"\x20\xE1\x40", 3};
+  static const optiga_metadata_item KEY_USE_SIGN = {(const uint8_t *)"\x10", 1};
+  static const optiga_metadata_item TYPE_PTFBIND = {(const uint8_t *)"\x22", 1};
+  optiga_metadata metadata = {0};
+
+  // Set metadata for device certificate.
+  memzero(&metadata, sizeof(metadata));
+  metadata.lcso = OPTIGA_LCS_OPERATIONAL;
+  metadata.change = OPTIGA_ACCESS_NEVER;
+  metadata.read = OPTIGA_ACCESS_ALWAYS;
+  metadata.execute = OPTIGA_ACCESS_ALWAYS;
+  if (!set_metadata(OID_CERT_DEV, &metadata)) {
+    return;
+  }
+
+  // Set metadata for FIDO attestation certificate.
+  memzero(&metadata, sizeof(metadata));
+  metadata.lcso = OPTIGA_LCS_OPERATIONAL;
+  metadata.change = OPTIGA_ACCESS_NEVER;
+  metadata.read = OPTIGA_ACCESS_ALWAYS;
+  metadata.execute = OPTIGA_ACCESS_ALWAYS;
+  if (!set_metadata(OID_CERT_FIDO, &metadata)) {
+    return;
+  }
+
+  // Set metadata for device private key.
+  memzero(&metadata, sizeof(metadata));
+  metadata.lcso = OPTIGA_LCS_OPERATIONAL;
+  metadata.change = OPTIGA_ACCESS_NEVER;
+  metadata.read = OPTIGA_ACCESS_NEVER;
+  metadata.execute = ACCESS_PAIRED;
+  metadata.key_usage = KEY_USE_SIGN;
+  if (!set_metadata(OID_KEY_DEV, &metadata)) {
+    return;
+  }
+
+  // Set metadata for FIDO attestation private key.
+  memzero(&metadata, sizeof(metadata));
+  metadata.lcso = OPTIGA_LCS_OPERATIONAL;
+  metadata.change = OPTIGA_ACCESS_NEVER;
+  metadata.read = OPTIGA_ACCESS_NEVER;
+  metadata.execute = ACCESS_PAIRED;
+  metadata.key_usage = KEY_USE_SIGN;
+  if (!set_metadata(OID_KEY_FIDO, &metadata)) {
+    return;
+  }
+
+  // Set metadata for pairing key.
+  memzero(&metadata, sizeof(metadata));
+  metadata.lcso = OPTIGA_LCS_OPERATIONAL;
+  metadata.change = OPTIGA_ACCESS_NEVER;
+  metadata.read = OPTIGA_ACCESS_NEVER;
+  metadata.execute = OPTIGA_ACCESS_ALWAYS;
+  metadata.data_type = TYPE_PTFBIND;
+  if (!set_metadata(OID_KEY_PAIRING, &metadata)) {
+    return;
+  }
+
+  vcp_println("OK");
+}
+
+static secbool is_optiga_locked(void) {
```

Comment:
I'm wondering what the use case for this function is. Shouldn't this function also check that the slots are configured correctly?

Reply:
The correct configuration is checked in the `LOCK` command. This is needed for the `VARIANT` command.

Reply:
Why do you need to check if the slots are locked in the `VARIANT` command? What informational value does it have for you if you don't know if the slots are configured correctly? Is that just a check to verify that the `LOCK` command was called before?

Reply:
> Why do you need to check if the slots are locked in the `VARIANT` command?

@matejcik or @Hannsek can probably provide more insight into the production process and rationale behind this.

> What informational value does it have for you if you don't know if the slots are configured correctly? Is that just a check to verify that the `LOCK` command was called before?

Yes.

Reply:
> Why do you need to check if the slots are locked in the `VARIANT` command? What informational value does it have for you if you don't know if the slots are configured correctly? Is that just a check to verify that the `LOCK` command was called before?

`LOCK` is called in the first phase of manufacturing setup. We assume that it checks its own success, i.e., that there is no need for a separate "verify metadata" step -- there is no service the host provides just by "asking again".

`VARIANT` is called in a subsequent phase by a different person on a different station at a different time. At that point, we need to ensure that a device has already gone through the first phase, which we detect by checking that `LOCK` has already been called (as the last step of phase 1)